### DISCUSSION TITLE
Scaling Puma to 500 requests per second w/ slow request handling

### DISCUSSION THREAD

**Discussion by jclusso** [puma/puma#3136] [Q&A]
I'll break this down into two parts because one is more configuration specific and the other is about handling slow requests?

### Configuration Question
From everything I've read I'm baffled by how Puma is supposed to be configured outside of Heroku or the like. All documentation and anything written recommends 1-1.5x workers per CPU and 5-6 threads. I can't figure out how this is optimal or scales when you start using EC2 instances.

For example, we have EC2 m6i.xlarge instances with 4 vCPU and 16GB and if I run them with 6 workers and 6 threads the servers will idle with 90% of their resources unused. Currently we're running them with 24 workers and and 25 threads and the CPU is idling at around 6-10% and the memory is around 60%. 

What I missing about how to configure Puma?

### Slow Request Handling
We have an API endpoint that has a default timeout of 5s and a maximum of 30s (user configurable). When a request comes in we queue a Sidekiq worker which will run for minutes if necessary to complete the process, but usually a whole request can be completed in an average of 400-500ms (fastest around 125ms). Our controller for this action is very simple. It validates the API request, queues a worker and then uses a mix of Redis blocking (first 5s) and polling (every second after) until the worker puts the full JSON response into Redis. 

Currently, we've implemented this Redis blocking and polling in a Lua script running in NGINX which offloads all of the slow request handling from Puma onto NGINX. Instead of the controller action polling it just responds to our Lua script which handles this. While this scales well, it's virtually a black box for monitoring and requires duplicated logic inside Lua (we want Lua failures to safely fallback and still work in Rails). I'd like to eliminate this complexity in our application and have Puma handle it, but I'm not sure Puma is capable of handling this task. 

Appreciative to anyone with more insight and feedback into how this should be optimally handled.

**Comment by nateberkopec** [comment]
> f I run them with 6 workers and 6 threads the servers will idle with 90% of their resources unused

You need more load. Puma can clearly handle more load in your setup. If you're seeing 10% CPU utilization but request queue times are above 1 second, then you must have a problem somewhere else in the setup.

>  I'd like to eliminate this complexity in our application and have Puma handle it, but I'm not sure Puma is capable of handling this task.

The solution is to not block. You should be immediately returning a 200 to the client and longpolling for the completed response, or use WebSockets. The current experience is suboptimal for the client, because they're simply on hold with no response while the server waits.

**Reply by jclusso** [reply]
> You need more load. Puma can clearly handle more load in your setup. If you're seeing 10% CPU utilization but request queue times are above 1 second, then you must have a problem somewhere else in the setup.

I'm going to work on testing this more. Any idea why we don't see our servers completely melting down or performance being completely obliterated when we have our settings so grossly over your recomendations? I've read all of your posts about tuning Puma, and I'm just baffled by the fact that I'm not seeing any signs that we're overloading our servers from a CPU, memory or response time standpoint. We additionally monitor queue time which averages around 4-5ms.

> The solution is to not block. You should be immediately returning a 200 to the client and longpolling for the completed response, or use WebSockets. The current experience is suboptimal for the client, because they're simply on hold with no response while the server waits.

This is only for a single API endpoint which our customers time and time again have told us they prefer a blocking request, rather than having to build retry logic. We currently allow them to pass a `timeout` parameter from 2-30s. I'm not aware of anyway we could use longpolling or WebSockets to do this unless there is something I am unaware of? 

Appreciate the advice!

**Reply by nateberkopec** [reply]
> Any idea why we don't see our servers completely melting down or performance being completely obliterated when we have our settings so grossly over your recomendations?

You _will_ get obliterated if you actually get some load (or if you reduce the server count enough). With too many processes and too many threads, request queue times will increase far more drastically over the recommended setup, because your workers will be ingesting more requests than they can actually process concurrently. With the recommended setup, workers will not ingest more load they can handle, which means that under 100% utilization request queue times and latency stays lower.

Scale down the number of servers you're using until you see request queue times start to increase. How are you managing autoscaling upward? What's the signal you use to trigger it?

**Reply by nateberkopec** [reply]
One thing I forgot - what percentage of your total request volume is this blocking/I/O wait endpoint? 

It's possible that if this is a large % of your traffic, that a higher thread count could be appropriate for you. If you're overall spending more than 50% of your time in I/O wait, a thread count of 10 would be good. If it's closer to 90%, even numbers as high as 25 could be optimal.

**Reply by jclusso** [reply]
Ok, that was what I was thinking might happen and is something we obviously don’t want to have happen. 

We don’t do auto scaling. Our load is fairly consistent and we try and keep a significant amount of head room. 

I’d say the bulk of our requests, probably 90% could be this endpoint. Right now with our Lua handler, these requests complete in 40-50ms and without would be 400-500ms. 

So if I understand you correctly, with 4 CPU it wouldn’t be unreasonable to do 6 workers and 25 threads assuming 90% are these requests. 

Any other suggestions in regards to eliminating this Lua handler we currently have? 

Thanks!

**Reply by nateberkopec** [reply]
> Any other suggestions in regards to eliminating this Lua handler we currently have?

Not really. If you're going to block the request, I don't see the point of involving Sidekiq or Lua. Just do it in the web process.

**Reply by jclusso** [reply]
The reason we need to use Sidekiq is because we want to wait nearly 5 minutes to honor timeouts of external servers according to RFCs. If the timeout of 5s default or 30s max are met, we return a try again response. This allows the worker to keep going and eventually get a response for the user.

**Reply by jclusso** [reply]
I've done more testing and I'm totally confused. When I had our servers configured with 32 workers and 64 threads I benchmarked up to 500 requests per second with a median response time of around 400-500ms. I reduced that to 6 workers and 25 threads and tried doing 100 requests per second and I immediately saw response times spike, puma pool capacity drop from 150 to 87 (average over 8 servers) and puma queue time spike up to a  850ms. Note, this is with our Lua handling disabled.

And I forgot, memory and CPU don't even appear to have moved much during that testing. CPU average was around 6% and memory is around 26%.

**Reply by nateberkopec** [reply]
It sounds like you just have a very unusual IO-heavy workload. Set worker count equal to CPU count, then tune thread count upward until request queue time (this is not the same as puma queue time, you must measure it externally) exceeds whatever value you deem acceptable (10-20% of total time usually).

**Reply by jclusso** [reply]
Got it, I'll try that and see what results I get. This controller action queues a worker and polls redis for completetion. When I said puma queue time, I was referring to our calculation from middleware we have. Is this what you mean by externally?

```ruby
class PumaLatencyMiddleware

  def initialize(app)
    @app = app
  end

  def call(env)
    start   = env['HTTP_X_REQUEST_START']&.sub('t=', '').to_f * 1000
    wait    = env['puma.request_body_wait']
    current = Time.now.to_f * 1000

    StatsD.histogram(
      'puma.queue_time',
      (current - wait - start).to_i
    ) unless start == 0

    @app.call(env)
  end

end

**Reply by nateberkopec** [reply]
Yup, looks good. I actually never thought of removing the request_body_wait before, but it makes sense.

**Reply by jclusso** [reply]
Here is my latest test results when running 100 requests per second with different configurations. I figured I'd try and do two comparisons that equate to the same total number of threads. Do you have any idea what would cause this because clearly the higher worker count with lower thread count drastically out performs. 

## Test 1
**Workers:** 4
**Threads:** 156

Within the first 5-10 seconds I had to terminate the test because the numbers spiked and this is a production stack. Additionally, it's worth noting I tried this same test with various lower thread counts like 10 and 25 with similiar results.

**Queue Time:** 15ms (i've seen this spike up to 800ms in various tests but I stopped it quick since we're in production)
**Median Response Time:** 2276ms (would've climbed to 10k ms)
**Puma Pool Running:** 625
**Puma Pool Capacity:** ~479
**Puma Pool Backlog:** 0
**Memory**: ~25%
**CPU:** <10%

## Test 2
**Workers:** 25
**Threads:** 25

I ran the test for 300 seconds. Over the entire test we had the following statistics. Additionally, since this test went well I ran the same thing for 250 requets per second and the results were nearly the same for all numbers. 

**Queue Time:** 3-4ms (5ms at 250/s)
**Median Response Time:** 376ms (391ms at 250/s)
**Puma Pool Running:** 625
**Puma Pool Capacity:** ~615 (560 at 250/s)
**Puma Pool Backlog:** 0
**Memory**: ~60%
**CPU:** <10% (~10-12% at 250/s)

**Reply by dentarg** [reply]
Isn't that the expected results? Matches everything Nate has stated about Puma and MRI/CRuby (the threads in Test 1 has to wait on the GIL?)

**Reply by jclusso** [reply]
> Isn't that the expected results? Matches everything Nate has stated about Puma and MRI/CRuby (the threads in Test 1 has to wait on the GIL?)

That doesn’t really explain how I’m running 25 workers instead of the suggested 4 and getting significantly better results.

**Reply by jclusso** [reply]
Chiming back in on this thread since it's been a while. Any further suggestions based on my most recent tests that completely violate the recomendation of 4 workers yet show significantly better performance?

**Reply by dentarg** [reply]
Haven't we established that the recommendation of 4 workers is for an app that perform compute work in its routes but yours are almost exclusively I/O? I'm not sure it is fair to judge Puma based on your specific application.

It would be great if you could benchmark an app that you can share, so others could try to reproduce/confirm any numbers presented.

**Reply by jclusso** [reply]
@dentarg I was thought that scaling the threads was where I should go and I should really stick to the 4 workers part. Maybe I misunderstood. Let me see if I can create a very barebones app to share.
